### PR TITLE
QUICKSTEP-20: Added ExecutionGenerator support for Window Aggregation

### PR DESCRIPTION
**Pull request by shixuan-fan**
- Created `WindowAggregationOperationState` and `WindowAggregationOperator`.
- The execution generator could form an execution plan for window aggregation. Window aggregation operator takes the result of sorting as input, and output a relation with one extra attribute which is the result of window aggregation.
- `WindowAggregationOperator::execute()` is a dummy method, it will report that window aggregation is not supported, and return an empty table based on the expected output relation.

### CODE REVIEW

File: query_optimizer/ExecutionGenerator.cpp [hunk 1]
```diff
@@ -1639,5 +1641,101 @@ void ExecutionGenerator::convertTableGenerator(
...
+
+  // Get window aggregate function expression.
+  const E::AliasPtr &named_window_aggregate_expression =
+      physical_plan->window_aggregate_expression();
+  const E::WindowAggregateFunctionPtr &window_aggregate_function =
+      std::static_pointer_cast<const E::WindowAggregateFunction>(
+          named_window_aggregate_expression->expression());
+
+  // Set the AggregateFunction.
+  window_aggr_state_proto->mutable_function()->CopyFrom(
```
Comment (zuyu) [inline query_optimizer/ExecutionGenerator.cpp]: Please use `MergeFrom`, as we don't need to clean up.

File: query_optimizer/ExecutionGenerator.cpp [hunk 2]
```diff
@@ -1639,5 +1641,101 @@ void ExecutionGenerator::convertTableGenerator(
...
+      window_aggregate_function->window_aggregate().getProto());
+
+  // Set the arguments.
+  for (const E::ScalarPtr &argument : window_aggregate_function->arguments()) {
+    unique_ptr<const Scalar> concretized_argument(argument->concretize(attribute_substitution_map_));
+    window_aggr_state_proto->add_arguments()->CopyFrom(concretized_argument->getProto());
```
Comment (zuyu) [inline query_optimizer/ExecutionGenerator.cpp]: Ditto for `MergeFrom`.

File: query_optimizer/ExecutionGenerator.cpp [hunk 3]
```diff
@@ -1639,5 +1641,101 @@ void ExecutionGenerator::convertTableGenerator(
...
+  }
+
+  // Set partition keys.
+  const E::WindowInfo window_info = window_aggregate_function->window_info();
+  for (const E::ScalarPtr &partition_by_attribute
+      : window_info.partition_by_attributes) {
+    unique_ptr<const Scalar> concretized_partition_by_attribute(
+        partition_by_attribute->concretize(attribute_substitution_map_));
+    window_aggr_state_proto->add_partition_by_attributes()
+        ->CopyFrom(concretized_partition_by_attribute->getProto());
```
Comment (zuyu) [inline query_optimizer/ExecutionGenerator.cpp]: Ditto for `MergeFrom`.

File: query_optimizer/ExecutionGenerator.cpp [hunk 4]
```diff
@@ -1639,5 +1641,101 @@ void ExecutionGenerator::convertTableGenerator(
...
+      window_aggregate_function->window_aggregate().getProto());
+
+  // Set the arguments.
+  for (const E::ScalarPtr &argument : window_aggregate_function->arguments()) {
+    unique_ptr<const Scalar> concretized_argument(argument->concretize(attribute_substitution_map_));
+    window_aggr_state_proto->add_arguments()->CopyFrom(concretized_argument->getProto());
+  }
+
+  // Set partition keys.
+  const E::WindowInfo window_info = window_aggregate_function->window_info();
```
Comment (zuyu) [inline query_optimizer/ExecutionGenerator.cpp]: I was wondering if we could use reference `&` here.

File: storage/WindowAggregationOperationState.proto [hunk 5]
```diff
@@ -0,0 +1,33 @@
...
+import "expressions/Expressions.proto";
+
+message WindowAggregationOperationState {
+  required int32 relation_id = 1;
+  required AggregateFunction function = 2;
+  repeated Scalar arguments = 3;
+  repeated Scalar partition_by_attributes = 4;
+  required bool is_row = 5;
+  required int64 num_preceding = 6;
+  required int64 num_following = 7;
```
Comment (zuyu) [inline storage/WindowAggregationOperationState.proto]: And comments for `num_preceding` and `num_following` regarding the usage of `-1`.

File: query_optimizer/ExecutionGenerator.cpp [hunk 6]
```diff
@@ -1639,5 +1641,101 @@ void ExecutionGenerator::convertTableGenerator(
...
+    //   1. If ORDER BY key is specified, use cumulative aggregation:
+    //      ROWS BETWEEN UNBOUNDED PRECEDING AND CURRENT ROW.
+    //   2. If ORDER BY key is not specified either, use the whole partition:
+    //      ROWS BETWEEN UNBOUNDED PRECEDING AND UNBOUNDED FOLLOWING.
+    window_aggr_state_proto->set_is_row(true);        /* frame mode: ROWS */
+    window_aggr_state_proto->set_num_preceding(-1);    /* UNBOUNDED PRECEDING */
+    if (window_info.order_by_attributes.empty()) {
+      window_aggr_state_proto->set_num_following(-1);  /* UNBOUNDED FOLLOWING */
+    } else {
+      window_aggr_state_proto->set_num_following(0);   /* CURRENT ROW */
```
Comment (zuyu) [inline query_optimizer/ExecutionGenerator.cpp]: Code style issue:

For above four lines w/ comments, we have two ways:

Either `window_aggr_state_proto->set_num_following(0 /* Current row */);`

Or `window_aggr_state_proto->set_num_following(0);  // Current row.`

And I prefer the second approach.

File: query_optimizer/ExecutionGenerator.cpp [hunk 7]
```diff
@@ -1639,5 +1641,101 @@ void ExecutionGenerator::convertTableGenerator(
...
+    }
+  } else {
+    const E::WindowFrameInfo* window_frame_info = window_info.frame_info;
```
Comment (zuyu) [inline query_optimizer/ExecutionGenerator.cpp]: Code style: `const E::WindowFrameInfo *window_frame_info`.

File: query_optimizer/cost_model/SimpleCostModel.hpp [hunk 8]
```diff
@@ -88,6 +89,10 @@ class SimpleCostModel : public CostModel {
   std::size_t estimateCardinalityForAggregate(
       const physical::AggregatePtr &physical_plan);
 
+  // Return the estimated cardinality of the input plan
```
Comment (zuyu) [inline query_optimizer/cost_model/SimpleCostModel.hpp]: Code style: Add the period `.` in the end.

File: storage/tests/WindowAggregationOperationState_unittest.cpp [hunk 9]
```diff
@@ -0,0 +1,91 @@
...
+
+using std::unique_ptr;
+
+namespace quickstep {
+
+class WindowAggregationOperationStateProtoTest : public ::testing::Test {
+ protected:
+  static const relation_id kInvalidTableId = 100;
+  static const std::int64_t kInvalidNum = -10;
+  static const std::int64_t kValidNum = 10;
```
Comment (zuyu) [inline storage/tests/WindowAggregationOperationState_unittest.cpp]: Optionally, like [here](https://github.com/apache/incubator-quickstep/blob/master/relational_operators/tests/SortRunGenerationOperator_unittest.cpp#L80) we could move these three values in an anonymous namespace:

```
namespace {
constexpr relation_id kInvalidTableId = 100;
constexpr std::int64_t kInvalidNum = -10;
constexpr std::int64_t kValidNum = 10;
}  // namespace
```

File: storage/tests/WindowAggregationOperationState_unittest.cpp [hunk 10]
```diff
@@ -0,0 +1,91 @@
...
+  proto.set_relation_id(rel_id_);
+  proto.mutable_function()->set_aggregation_id(serialization::AggregateFunction::AVG);
+  proto.set_is_row(true);
+  proto.set_num_preceding(kValidNum);
+  proto.set_num_following(kValidNum);
+  EXPECT_TRUE(WindowAggregationOperationState::ProtoIsValid(proto, *database_.get()));
+}
+
+}  // namespace quickstep
+
```
Comment (zuyu) [inline storage/tests/WindowAggregationOperationState_unittest.cpp]: Please remove this empty line.

**Comment by shixuan-fan** [comment]
@zuyu  Thanks a lot for your review. Addressed them in the latest commit :).

**Pull request merged**